staging/src/k8s.io/apiserver/pkg/endpoints/filters/traces.go

```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"net/http"
+
+	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
+	"go.opentelemetry.io/otel/trace"
+
+	"k8s.io/component-base/traces"
+)
+
+// WithTracing adds tracing to requests if the incoming request is sampled
+func WithTracing(handler http.Handler, tp *trace.TracerProvider) http.Handler {
+	opts := []otelhttp.Option{
+		otelhttp.WithPropagators(traces.Propagators()),
+	}
+	if tp != nil {
```

Comment:
if tp is nil, should we just do `return handler`?

Reply:
A tracer provider is required to export traces, but even without a tracer provider we can still propagate context.  That would allow, for example, linking etcd traces with API requests even if trace collection is not configured for the api server.  I added a comment linking to otel documentation.

Reply:
interesting. is there any reason someone would not want that once this feature has graduated to GA and can't be disabled via gate?

Reply:
Generally not, although there may be some users that want a different set of propagators, like the zipkin B3 propagators.

There is a related use-case in which the kube endpoint is public, and you don't trust the context from incoming requests.  For that case we wouldn't want to remove propagators, though.  We would want to use the [WithPublicEndpoint](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/net/http/otelhttp/config.go#L103) option on the handler to change the relationships with incoming SpanContext from being a Child to a Link.

Reply:
> There is a related use-case in which the kube endpoint is public, and you don't trust the context from incoming requests. For that case we wouldn't want to remove propagators, though. We would want to use the WithPublicEndpoint option on the handler to change the relationships with incoming SpanContext from being a Child to a Link.

Given the filter is added so it runs before authn/authz, including on generally-accessible endpoints like /healthz, isn't that effectively already possible?

Reply:
We should probably restrict tracing only to authenticated endpoints (and I moved the tracing filter so it only runs on authenticated requests).  Is there a good list somewhere of endpoints that are authenticated?  Or should I just start with  only tracing requests to `/api`?

Reply:
> I moved the tracing filter so it only runs on authenticated requests

for most clusters that permit anonymous requests to some endpoints, that filter can just decorate a request with no credentials with the "system:anonymous" user :)

what are the implications of an anonymous user (or authenticated-but-untrusted user, which is potentially all clusters) having control over Child relationships vs Link relationships?

Reply:
Ah, the important distinction is actually that a Link ignores the incoming span context (other than linking it), and doesn't use it for determining whether to sample or not.  If an anonymous user can create children, it can make all requests sampled.

In general, we want to be able to create children, rather than links, because debugging re-entrant API calls from webhook admission controllers is one of the core use-cases we want to solve.  Ideally we can do that without allowing anyone to generate a bunch of spans.  How do we solve that problem with audit logs?  Do we still audit unauthenticated requests?

Reply:
We do audit unauthenticated requests, though most unauthenticated requests would get stopped at the authorization layer.

Reply:
Putting this after the authorization check would mean you could only influence tracing for endpoints/verbs you were actually authorized to request.

I still don't have a clear picture of what "influencing tracing" means for an unauthenticated or authenticated-but-untrusted user.

If all users (anonymous or otherwise) are authorized to `GET /api/v1/namespaces/kube-public/configmaps/cluster-info` (as they are on kubeadm clusters), what info/headers/params could they pass in that would affect totally unrelated traces?

Reply:
It wouldn't affect completely unrelated traces, except that you might be able to DOS your tracing backend.

Reply:
> It wouldn't affect completely unrelated traces

does the info passed in via headers determine what is unrelated? is there anything an anonymous/untrusted requester could pass in in a header that would make tracing think this request was related to another request incorrectly?

Reply:
they'd need access to that other request though.. otherwise how would they generate the trace id they shouldn't have?

Reply:
It is possible to append spans to an existing trace, if you know the trace ID, and one of the span IDs.  But spans are independent, so you wouldn't be able to alter any existing spans in the trace.

Reply:
> It wouldn't affect completely unrelated traces, except that you might be able to DOS your tracing backend. Is that something I should be worrying about?

Um... maybe? I see pros/cons the further back in the filter chain we put this:

* after authentication
  * pro: prevents requests with bad credentials from initiating traces (but not valid credentials or anonymous requests)
  * con: means tracing doesn't have visibility to costs of remote authentication methods like webhook authentication
* after priority-and-fairness
  * pro: prevents requests rejected because they exceed allowed priority-and-fairness threshholds from initiating traces
  * con: means tracing doesn't have visibility to requests delayed/rejected for this reason
* after authorization
  * pro: prevents unauthorized requests from initiating traces
  * con: means tracing doesn't have visibility to costs of remote authorization methods like webhook authorization

Even it we put it after all of those, it still sounds like an authorized request could saturate the tracing backend. What's the failure mode in that scenario?

> they'd need access to that other request though.. otherwise how would they generate the trace id they shouldn't have?

if the trace id is considered sufficiently obscure/secure to prevent abuse, that seems positive, and might limit them to influencing traces related to their own requests.

does a new trace id established for a request get returned to the user making that request? are there bad implications of them passing that into subsequent requests and establishing bogus child relationships (like the trace backend retaining the parent traces longer, running out of memory, etc)?

Reply:
The failure mode for saturation depends on the backend, but probably involves dropping trace data.

If a user sends a request without a trace id, it isn't ever returned to them.  But they can start a new trace (by creating a span without a parent themselves), and send it with the request to the apiserver.

The answer again will probably depend on the backend, but I dug into jaeger, and it looks like jaeger has a max duration for a trace (determined by the `max-span-age` flag) and it will ignore spans older than that age.  That would prevent creating traces that are ~infinitely long.  I haven't heard of that causing OOM, or anything like that.

Reply:
Having visibility into the costs of remote authentication would be really nice...

Reply:
> Having visibility into the costs of remote authentication would be really nice...

agreed... mostly trying to get my head around the risks this exposes. @lavalamp, you've thought about the abuse vectors more w.r.t. priority and fairness... do you have thoughts on this?

Reply:
I went ahead and added WithPublicEndpoint, which makes it ignore the incoming span context.  If that is acceptable, we can start with that and continue the discussion in an issue.

Reply:
sgtm, mind including a link to this thread (or hoisting relevant content) in the issue?